lms/views/decorators/h_api.py

```diff
@@ -0,0 +1,126 @@
+"""View decorators for working with the h API."""
+
+import functools
+
+from pyramid.httpexceptions import HTTPBadRequest
+
+from lms.services import HAPINotFoundError
+
+
+__all__ = ["upsert_h_user", "upsert_course_group", "add_user_to_group"]
+
+
+def upsert_h_user(wrapped):
+    """
+    Create or update the Hypothesis user for the request's LTI user.
+
+    Update the h user's information from LTI data. If the user doesn't exist
+    yet, call the h API to create one.
+
+    Assumes that it's only used on LTI launch views.
+    """
+
+    @functools.wraps(wrapped)
+    def wrapper(context, request):
+        if not context.provisioning_enabled:
+            return wrapped(context, request)
+
+        hapi_svc = request.find_service(name="hapi")
+
+        user_data = {
+            "username": context.h_username,
+            "display_name": context.h_display_name,
+            "authority": request.registry.settings["h_authority"],
+            "identities": [
+                {
+                    "provider": context.h_provider,
+                    "provider_unique_id": context.h_provider_unique_id,
+                }
+            ],
+        }
+
+        try:
+            hapi_svc.patch(
+                f"users/{context.h_username}", {"display_name": context.h_display_name}
+            )
+        except HAPINotFoundError:
+            # Call the h API to create the user in h if it doesn't exist already.
+            hapi_svc.post("users", user_data)
+
+        return wrapped(context, request)
+
+    return wrapper
+
+
+def upsert_course_group(wrapped):
+    """
+    Create or update the Hypothesis group for the request's LTI course.
+
+    Call the h API to create a group for the LTI course, if one doesn't exist
+    already.
+
+    Groups can only be created if the LTI user is allowed to create Hypothesis
+    groups (for example instructors are allowed to create groups). If the group
+    for the course hasn't been created yet, and if the user isn't allowed to
+    create groups (e.g. if they're just a student) then show an error page
+    instead of continuing with the LTI launch.
+
+    Assumes that it's only used on LTI launch views.
+    """
+
+    @functools.wraps(wrapped)
+    def wrapper(context, request):
+        if not context.provisioning_enabled:
```

Comment:
Because I'm seeing this guard in every one of these wrappers, it makes me wonder if there is any way to factor it out...perhaps not. Just thought I'd mention it. An argument against doing anything would be that the code is quite easy to read (and not magical) as-is.

Reply:
It was supposed to be promptly deleted as soon as provisioning was enabled for everyone, and but here we are six months later still supporting LMS users who have provisioning turned off. We should really turn it on for everyone...